cvmfs/catalog_balancer.h

```diff
@@ -0,0 +1,146 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_CATALOG_BALANCER_H_
+#define CVMFS_CATALOG_BALANCER_H_
+
+#include <string>
+#include <vector>
+
+#include "catalog_mgr.h"
+#include "directory_entry.h"
+
+
+namespace catalog {
+
+/**
+ * This class is in charge of "balancing" a catalog embedded in a
+ * WritableCatalogManager. The process of balancing consists of keeping the
+ * number of entries in each catalog between a maximum and minimum threshold so
+ * that all of them can be easily manipulable. This way there won't be catalogs
+ * that take a lot of time being downloaded, or their SQL operations will take
+ * a reasonable amount of time.
+ *
+ * The CatalogBalancer uses the following WritableCatalogManager attributes:
+ * - max_weight_: maximum number of entries in a catalog
+ * - min_weight_: minimum number of entries in a catalog
+ * - balance_weight_: maximum number of entries during the balancing process
+ *
+ * The balancing process is done traversing the entire catalog tree in
+ * postorder, so that the deepest catalogs are analyzed before their parents.
+ * Each catalog is then analyzed individually, existing three different
+ * scenarios:
+ *
+ * a) The number of entries of the catalog is between max_weight_ and
+ * min_weight_: nothing happens, the catalog remains untouched.
+ *
+ * b) The number of entries of the catalog is greater than max_weight_: the
+ * catalog gets split in smaller catalogs. One of the new generated catalogs
+ * will be mounted in the original mount point.
+ *
+ * c) The number of entries of the catalog is lesser than min_weight_: the
+ * catalog gets merged with its father (except the root catalog, obviously).
+ */
+template <class CatalogMgrT>
+class CatalogBalancer {
+ public:
+  typedef typename CatalogMgrT::catalog_t catalog_t;
```

Comment:
I finally used the CatalogMgrT::catalog_t for this, so it shouldn't be removed.
This is simply to make the code clearer, instead of putting so many typenames